deep_quoridor/src/quoridor_env.py

```diff
@@ -80,12 +88,161 @@ def step(self, action):
             self._place_wall(agent, (row, col), action_type - 1)
 
         if self._check_win(agent):
-            self.terminations = {a: True for a in self.possible_agents}
+            self.terminations = {a: True for a in self.agents}
             self.rewards[agent] = 1
             self.rewards[self._opponent(agent)] = -1
 
+        # TODO: Confirm if this is needed and if it's doing anything
+        self._accumulate_rewards()
+
         self._next_player()
 
+    def observe(self, agent_id):
+        """
+        Returns the observation and action mask in a dict, like so:
+        {
+            "observation": observation,
+            "action_mask": action_mask
+        }
+        """
+        return {
+            "observation": self._get_observation(agent_id),
+            "action_mask": self._get_action_mask(agent_id),
+        }
+
+    def _get_observation(self, agent_id):
+        # TODO: Do we need to make copies of the state or can we return references directly?
+
+        # Calculate board from self.positions
+        # NOTE: The board uses 1 to indicate where the agent is and 2 to indicate where the opponent is
+        # Obviously, this will be different for each player
+        board = np.zeros((self.board_size, self.board_size), dtype=np.int8)
+        board[self.positions[agent_id]] = 1
+        board[self.positions[self._opponent(agent_id)]] = 2
+
+        # Make a copy of walls
+        walls = self.walls.copy()
+
+        return {
+            "board": board,
+            "walls": walls,
+            "my_walls_remaining": self.walls_remaining[agent_id],
+            "opponent_walls_remaining": self.walls_remaining[self._opponent(agent_id)],
+        }
+
+    def _get_action_mask(self, agent_id):
+        # Start with an empty mask (nothing possible)
+        mask = np.zeros((self.board_size**2 + (self.wall_size**2) * 2,), dtype=np.int8)
+
+        # Calculate valid "moves" (as opposed to wall placements)
+        # Start with the four basic directions
+        for row, col in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
+            old_row, old_col = self.positions[agent_id]
+            new_row, new_col = old_row + row, old_col + col
+
+            # Check if the new position is still inside the board
+            if not (0 <= new_row < self.board_size and 0 <= new_col < self.board_size):
+                continue
+
+            # Is there a wall in the way we're going?
+            blocked_by_wall = False
+            if row == 0:  # Moving vertically - check horizontal walls
```

Comment:
The comment should be the opposite, since the row is not changing, right?
```suggestion
            if row == 0:  # Moving horizontally - check vertical walls
```